### PR TITLE
Replace all crypto libraries with Botan

### PR DESCRIPTION
**Pull request by droidmonkey**
Selected the [Botan crypto library](https://github.com/randombit/botan) due to feature list, maintainer support, availabilty across all deployment platforms, and ease of use. Also evaluated Crypto++ as a viable candidate, but the additional features of Botan (PKCS#11, TPM, etc) won out.

The random number generator received a backend upgrade. Botan prefers hardware-based RNG's and will provide one if available. This is transparent to KeePassXC and a significant improvement over gcrypt.

Replaced Argon2 library with built-in Botan implementation that supports i, d, and id. This requires Botan 2.11.0 or higher. Also simplified the parameter test across KDF's.

Aligned SymmetricCipher parameters with available modes. All encrypt and decrypt operations are done in-place instead of returning new objects. This allows use of secure vectors in the future with no additional overhead.

Took this opportunity to decouple KeeShare from SSH Agent. Removed leftover code from OpenSSHKey and consolidated the SSH Agent code into the same directory. Removed bcrypt and blowfish inserts since they are provided by Botan.

Additionally simplified KeeShare settings interface by removing raw certificate byte data from the user interface. KeeShare will be further refactored in a future PR.

_NOTE: This PR breaks backwards compatibility with KeeShare certificates due to different RSA key storage with Botan. As a result, new "own" certificates will need to be generated and trust re-established._

Removed YKChallengeResponseKeyCLI in favor of just using the original implementation with signal/slots.

Removed TestRandom stub since it was just faking random numbers and not actually using the backend. TestRandomGenerator now uses the actual RNG.

Greatly simplified Secret Service plugin's use of crypto functions with Botan.

--------

**This refactor directly supports the move the MSVC on Windows. It also supports the inclusion of PKCS#11 hardware tokens for database credentials and use of TPM for in-memory encryption.**

**Special thanks to @randombit for his work on Botan!**

--------

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tested on all platforms. Xenial and Bionic require use of PPA to bring Botan 2.11 or higher. This version requirement is mainly driven by the Argon2 support.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Refactor (significant modification to existing code)

### CODE REVIEW

**Comment by hifi** [comment]
I've been using this for a few days with KDBX4+AES-256+AES-KDF, no issues yet. SSH Agent can also handle unencrypted ED25519 and RSA keys correctly in daily use and my personal test database that decrypts and loads a bunch of different keys seems to work as well.

All non-GUI tests have passed on both x86_64 and AArch64 without any issues.

**Comment by yan12125** [comment]
This also works for ChaCha20+Argon2id and encrypted Ed25519 SSH keys. Fabulous work!

Looks like botan is not installed on CI machines yet?

File: src/keeshare/KeeShareSettings.cpp [hunk 1]
```diff
@@ -106,17 +70,25 @@ namespace KeeShareSettings
...
     }
 
     bool Certificate::operator==(const Certificate& other) const
     {
-        return key == other.key && signer == other.signer;
+        if (isNull() && other.isNull()) {
+            return true;
+        } else if (isNull() || other.isNull()) {
+            return false;
+        }
```
Comment (phoerious) [inline src/keeshare/KeeShareSettings.cpp:90]:
```suggestion
        if (isNull() || other.isNull()) {
            return isNull() == other.isNull();
        }
```

File: src/keeshare/KeeShareSettings.cpp [hunk 2]
```diff
@@ -157,15 +116,29 @@ namespace KeeShareSettings
...
     }
 
     bool Key::operator==(const Key& other) const
     {
-        return key == other.key;
+        if (isNull() && other.isNull()) {
+            return true;
+        } else if (isNull() || other.isNull()) {
+            return false;
+        }
```
Comment (phoerious) [inline src/keeshare/KeeShareSettings.cpp:140]: See above.

File: src/keeshare/KeeShareSettings.cpp [hunk 3]
```diff
@@ -157,15 +116,29 @@ namespace KeeShareSettings
...
             } else if (reader.name() == "Key") {
-                certificate.key = QByteArray::fromBase64(reader.readElementText().toLatin1());
+                auto rawKey = QByteArray::fromBase64(reader.readElementText().toLatin1());
+                if (!rawKey.isEmpty()) {
+                    try {
+                        Botan::DataSource_Memory dataSource(reinterpret_cast<const uint8_t*>(rawKey.constData()),
+                                                            rawKey.size());
+                        certificate.key.reset(Botan::PKCS8::load_key(dataSource).release());
+                    } catch (std::exception& e) {
+                        qWarning("KeeShare: Failed to deserialize key data: %s", e.what());
```
Comment (phoerious) [inline src/keeshare/KeeShareSettings.cpp:126]: This deserves proper error handling.

File: src/keeshare/KeeShareSettings.cpp [hunk 4]
```diff
@@ -16,77 +16,41 @@
...
+#include <botan/ber_dec.h>
+#include <botan/pk_keys.h>
+#include <botan/pkcs8.h>
+#include <botan/rsa.h>
 #include <functional>
 
 namespace KeeShareSettings
 {
     namespace
     {
```
Comment (phoerious) [inline src/keeshare/KeeShareSettings.cpp:43]: Unindent.

File: src/keeshare/KeeShareSettings.cpp [hunk 5]
```diff
@@ -175,45 +148,45 @@ namespace KeeShareSettings
...
         Key key;
-        key.key = QByteArray::fromBase64(reader.readElementText().toLatin1());
+        auto rawKey = QByteArray::fromBase64(reader.readElementText().toLatin1());
+        if (!rawKey.isEmpty()) {
+            try {
+                Botan::DataSource_Memory dataSource(reinterpret_cast<const uint8_t*>(rawKey.constData()),
+                                                    rawKey.size());
+                key.key.reset(Botan::PKCS8::load_key(dataSource).release());
+            } catch (std::exception& e) {
+                qWarning("KeeShare: Failed to deserialize key data: %s", e.what());
```
Comment (phoerious) [inline src/keeshare/KeeShareSettings.cpp:174]: Error handling.

File: src/keeshare/KeeShareSettings.h [hunk 6]
```diff
@@ -20,8 +20,13 @@
 
 #include <QMap>
 #include <QObject>
+#include <QSharedPointer>
+#include <QUuid>
 
-#include "crypto/ssh/OpenSSHKey.h"
+namespace Botan
+{
+    class Private_Key;
```
Comment (phoerious) [inline src/keeshare/KeeShareSettings.h:28]: Unindent.

File: src/keys/FileKey.cpp [hunk 7]
```diff
@@ -225,7 +212,7 @@ void FileKey::createXMLv2(QIODevice* device, int size)
         }
         w.writeCharacters(QChar(key[i]));
     }
-    sodium_memzero(key.data(), static_cast<std::size_t>(key.capacity()));
+    Botan::secure_scrub_memory(key.data(), static_cast<std::size_t>(key.capacity()));
```
Comment (phoerious) [inline src/keys/FileKey.cpp:215]: Make `key` a `Botan::secure_vector`, then you don't need to scrub it here.

Also, we may want to disable Botan's mlock allocator for ASAN builds. See https://github.com/randombit/botan/issues/626

File: src/keys/FileKey.cpp [hunk 8]
```diff
@@ -338,19 +325,19 @@ bool FileKey::loadXml(QIODevice* device, QString* errorMsg)
...
 
     bool ok = false;
     if (!xmlReader.error() && !keyFileData.data.isEmpty()) {
-        std::memcpy(m_key, keyFileData.data.data(), std::min(SHA256_SIZE, keyFileData.data.size()));
+        std::memcpy(m_key.data(), keyFileData.data.data(), std::min(SHA256_SIZE, keyFileData.data.size()));
         ok = true;
     }
 
-    sodium_memzero(keyFileData.data.data(), static_cast<std::size_t>(keyFileData.data.capacity()));
+    Botan::secure_scrub_memory(keyFileData.data.data(), static_cast<std::size_t>(keyFileData.data.capacity()));
```
Comment (phoerious) [inline src/keys/FileKey.cpp:340]: See above.

File: src/keys/FileKey.cpp [hunk 9]
```diff
@@ -373,8 +360,8 @@ bool FileKey::loadBinary(QIODevice* device)
         return false;
     }
 
-    std::memcpy(m_key, data.data(), std::min(SHA256_SIZE, data.size()));
-    sodium_memzero(data.data(), static_cast<std::size_t>(data.capacity()));
+    std::memcpy(m_key.data(), data.data(), std::min(SHA256_SIZE, data.size()));
+    Botan::secure_scrub_memory(data.data(), static_cast<std::size_t>(data.capacity()));
```
Comment (phoerious) [inline src/keys/FileKey.cpp:364]: See above.

File: src/keys/FileKey.cpp [hunk 10]
```diff
@@ -402,14 +389,14 @@ bool FileKey::loadHex(QIODevice* device)
     }
 
     QByteArray key = QByteArray::fromHex(data);
-    sodium_memzero(data.data(), static_cast<std::size_t>(data.capacity()));
+    Botan::secure_scrub_memory(data.data(), static_cast<std::size_t>(data.capacity()));
```
Comment (phoerious) [inline src/keys/FileKey.cpp:392]: Here also.

File: src/keys/FileKey.cpp [hunk 11]
```diff
@@ -402,14 +389,14 @@ bool FileKey::loadHex(QIODevice* device)
...
 
     if (key.size() != 32) {
         return false;
     }
 
-    std::memcpy(m_key, key.data(), std::min(SHA256_SIZE, key.size()));
-    sodium_memzero(key.data(), static_cast<std::size_t>(key.capacity()));
+    std::memcpy(m_key.data(), key.data(), std::min(SHA256_SIZE, key.size()));
+    Botan::secure_scrub_memory(key.data(), static_cast<std::size_t>(key.capacity()));
```
Comment (phoerious) [inline src/keys/FileKey.cpp:399]: Here.

File: src/keys/FileKey.cpp [hunk 12]
```diff
@@ -434,8 +421,8 @@ bool FileKey::loadHashed(QIODevice* device)
     } while (!buffer.isEmpty());
 
     auto result = cryptoHash.result();
-    std::memcpy(m_key, result.data(), std::min(SHA256_SIZE, result.size()));
-    sodium_memzero(result.data(), static_cast<std::size_t>(result.capacity()));
+    std::memcpy(m_key.data(), result.data(), std::min(SHA256_SIZE, result.size()));
+    Botan::secure_scrub_memory(result.data(), static_cast<std::size_t>(result.capacity()));
```
Comment (phoerious) [inline src/keys/FileKey.cpp:425]: And here.

File: src/keys/YkChallengeResponseKey.cpp [hunk 13]
```diff
@@ -43,35 +39,17 @@ YkChallengeResponseKey::YkChallengeResponseKey(YubiKeySlot keySlot)
...
-        if (m_key) {
-            gcry_free(m_key);
-        }
-        m_keySize = static_cast<std::size_t>(key.size());
-        m_key = static_cast<char*>(gcry_malloc_secure(m_keySize));
-        std::memcpy(m_key, key.data(), m_keySize);
-        sodium_memzero(key.data(), static_cast<std::size_t>(key.capacity()));
+        m_key.resize(response.size());
+        Botan::copy_mem(m_key.data(), response.constData(), response.size());
+        Botan::secure_scrub_memory(response.data(), response.capacity());
```
Comment (phoerious) [inline src/keys/YkChallengeResponseKey.cpp:52]: `response` should also be a `secure_vector`.

File: src/main.cpp [hunk 14]
```diff
@@ -96,6 +96,14 @@ int main(int argc, char** argv)
...
     }
 
+    // Show debug information and then exit
+    if (parser.isSet(debugInfoOption)) {
+        QTextStream out(stdout, QIODevice::WriteOnly);
+        QString debugInfo = Tools::debugInfo().append("\n").append(Crypto::debugInfo());
+        out << debugInfo << endl;
+        return EXIT_SUCCESS;
+    }
+
```
Comment (phoerious) [inline src/main.cpp:106]: Move to separate commit.

File: src/main.cpp [hunk 15]
```diff
@@ -128,14 +127,14 @@ int main(int argc, char** argv)
...
     }
 
-    // Displaying the debugging informations must be done after Crypto::init,
-    // to make sure we know which libgcrypt version is used.
-    if (parser.isSet(debugInfoOption)) {
-        QTextStream out(stdout, QIODevice::WriteOnly);
-        QString debugInfo = Tools::debugInfo().append("\n").append(Crypto::debugInfo());
-        out << debugInfo << endl;
-        return EXIT_SUCCESS;
-    }
```
Comment (phoerious) [inline src/main.cpp:138]: See above.

File: tests/TestSharing.h [hunk 16]
```diff
@@ -21,16 +21,16 @@
 #include <QObject>
 #include <QSharedPointer>
 
-class OpenSSHKey;
-
+namespace Botan
+{
+    class RSA_PrivateKey;
```
Comment (phoerious) [inline tests/TestSharing.h:26]: Unindent.

File: tests/TestSignature.cpp [hunk 17]
```diff
@@ -107,92 +111,95 @@ static const char* rsa_1_public = "-----BEGIN RSA PUBLIC KEY-----\n"
...
+QSharedPointer<Botan::Private_Key> loadPrivateKey(const QString& pem)
+{
+    try {
+        std::string label;
+        auto der = Botan::PEM_Code::decode(pem.toStdString(), label);
+        auto key = new Botan::RSA_PrivateKey(
+            Botan::AlgorithmIdentifier("RSA", Botan::AlgorithmIdentifier::USE_NULL_PARAM), der);
+        return QSharedPointer<Botan::Private_Key>(key);
+    } catch (std::exception& e) {
+        qWarning("Failed to load key: %s", e.what());
```
Comment (phoerious) [inline tests/TestSignature.cpp:123]: This is a test suite. You should QFAIL outa here.

File: tests/TestSignature.cpp [hunk 18]
```diff
@@ -107,92 +111,95 @@ static const char* rsa_1_public = "-----BEGIN RSA PUBLIC KEY-----\n"
...
+{
+    try {
+        std::string label;
+        auto der = Botan::PEM_Code::decode(pem.toStdString(), label);
+        auto key =
+            new Botan::RSA_PublicKey(Botan::AlgorithmIdentifier("RSA", Botan::AlgorithmIdentifier::USE_NULL_PARAM),
+                                     std::vector<uint8_t>(der.begin(), der.end()));
+        return QSharedPointer<Botan::Public_Key>(key);
+    } catch (std::exception& e) {
+        qWarning("Failed to load key: %s", e.what());
```
Comment (phoerious) [inline tests/TestSignature.cpp:138]: Same here.

**Review by phoerious — CHANGES_REQUESTED** [review]
Monster PR, but nice net code reduction.

File: src/crypto/Crypto.cpp [hunk 19]
```diff
@@ -19,326 +19,288 @@
...
+        QByteArray key = QByteArray::fromHex("000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F");
+        QByteArray iv = QByteArray::fromHex("00000000000000000000000000000000");
+        QByteArray plainText = QByteArray::fromHex("00112233445566778899AABBCCDDEEFF");
+        plainText.append(QByteArray::fromHex("00112233445566778899AABBCCDDEEFF"));
+        QByteArray cipherText = QByteArray::fromHex("8EA2B7CA516745BFEAFC49904B496089");
+        cipherText.append(QByteArray::fromHex("8EA2B7CA516745BFEAFC49904B496089"));
 
-    return true;
-}
+        std::unique_ptr<Botan::BlockCipher> cipher(Botan::BlockCipher::create("AES-256"));
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/crypto/Crypto.cpp:188]: I think this was temporary code that made it into production, good catch.

File: src/crypto/Crypto.cpp [hunk 20]
```diff
@@ -19,326 +19,288 @@
...
-        raiseError(aes256Encrypt.errorString());
-        return false;
-    }
-    if (encryptedText != cipherText) {
-        raiseError("AES-256 ECB encryption mismatch.");
-        return false;
-    }
+        Botan::secure_vector<uint8_t> out(plainText.size());
+        cipher->encrypt_n(
+            reinterpret_cast<const uint8_t*>(plainText.data()), out.data(), plainText.size() / cipher->block_size());
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/crypto/Crypto.cpp:193]: This is true

File: src/crypto/Crypto.cpp [hunk 21]
```diff
@@ -19,326 +19,288 @@
...
+            return false;
+        }
+        if (!salsa20Stream.process(data)) {
+            raiseError(salsa20Stream.errorString());
+            return false;
+        }
+        if (data != salsa20Cipher) {
+            raiseError("Salsa20 stream cipher mismatch.");
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/crypto/Crypto.cpp:272]: Stream ciphers are bidirectional by design you wouldn't technically be proving anything, but we can add the check.

File: src/crypto/CryptoHash.cpp [hunk 22]
```diff
@@ -82,39 +71,52 @@ void CryptoHash::addData(const QByteArray& data)
...
+            d->hmacFunction->update(reinterpret_cast<const uint8_t*>(data.data()), data.size());
+        } else if (d->hashFunction) {
+            d->hashFunction->update(reinterpret_cast<const uint8_t*>(data.data()), data.size());
+        }
+    } catch (std::exception& e) {
+        qWarning("CryptoHash::update failed to add data: %s", e.what());
+    }
 }
 
 void CryptoHash::setKey(const QByteArray& data)
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/crypto/CryptoHash.cpp:85]: That is coming in the next PR, want to make this work with current data container then transition to secure data container.

File: src/crypto/kdf/AesKdf.cpp [hunk 23]
```diff
@@ -83,43 +84,26 @@ bool AesKdf::transform(const QByteArray& raw, QByteArray& result) const
...
 
     quint64 rounds = 1000000;
     QElapsedTimer timer;
     timer.start();
 
-    if (!cipher.processInPlace(key, rounds)) {
+    QByteArray result;
+    if (!transformKeyRaw(key, seed, rounds, &result)) {
         return -1;
     }
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/crypto/kdf/AesKdf.cpp:108]: Yah we can add a "repeat" parameter to run the test multiple times and average out the result.

File: src/crypto/Crypto.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (phoerious) [reply src/crypto/Crypto.cpp:272]: You would prove that what comes out of the encryption pass is actually the result of a stream cipher. But, yeah, we check for the result already, so it would be more of a double check that the test ist correct. You may also leave it out.

File: src/crypto/CryptoHash.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/crypto/CryptoHash.cpp:85]: Also this container can easily add in memory encryption as a feature

File: src/crypto/Crypto.cpp [hunk 24]
```diff
@@ -19,326 +19,288 @@
...
+                          .arg(Botan::version_patch());
+            return false;
+        }
+        if (Botan::version_minor() < 11) {
+            m_error = QObject::tr("Botan library version must be at least 2.11.x, found %1.%2.%3")
+                          .arg(Botan::version_major())
+                          .arg(Botan::version_minor())
+                          .arg(Botan::version_patch());
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/crypto/Crypto.cpp:66]: agree, this was overkill

File: src/crypto/CryptoHash.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (phoerious) [reply src/crypto/CryptoHash.cpp:85]: Probably not even needed. Unless you have weird ulimits, secure_vector is mocked and should therefore not end up in swap files. We should rather change how we handle the inner stream, which we don't need to decrypt up front if we don't need it.

**Comment by ofer-dev** [comment]
Does this affect key derivation time? My understanding is the Botan currently do not provide simd(sse, avx/2/512) acceleration while libargon2 does.

**Comment by randombit** [comment]
As best I can tell SSE2/AVX2 does not actually help Argon2 performance that much, probably because most of the cost in in memory access. Comparing `argon2` command line util vs Botan, Argon2i, 3 passes, 1 thread

64M libargon2 163 ms Botan 134 ms
256M libargon2 670 ms Botan 560 ms
1024M libargon2 2783 ms Botan 2323 ms

This is on an i7-6700K running Linux.

Possibly I'm comparing the wrong libargon2? These numbers are with https://github.com/P-H-C/phc-winner-argon2

**Comment by droidmonkey** [comment]
> Possibly I'm comparing the wrong libargon2? These numbers are with [P-H-C/phc-winner-argon2](https://github.com/P-H-C/phc-winner-argon2)

This was the previous implementation of Argon2 that we were using and Botan has replaced in this PR.

**Comment by phoerious** [comment]
I guess the point is that you want your own Argon2 implementation to be as fast as possible so as to be able to perform as many rounds as possible, because an attacker most certainly will use the most optimised version out there. I really wouldn't use KeePass2's C# implementation for exactly that reason, which is orders of magnitude slower than what we have. Though I doubt that the difference between Botan and libargon2 is really significant, but I didn't benchmark it.

**Comment by droidmonkey** [comment]
Right, I meant hardware optimization in my comment above.

**Comment by phoerious** [comment]
It will be possible to some extent, but it shouldn't give you a 10x boost.

File: src/crypto/Crypto.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/crypto/Crypto.cpp:193]: I removed this entirely and replaced it with calling the AES KDF function defined in Symmetric Cipher instead.

File: src/crypto/Crypto.h [hunk 25]
```diff
@@ -20,31 +20,12 @@
...
-
-    static bool m_initialized;
-    static QString m_errorStr;
-    static QString m_backendVersion;
-};
+    bool init();
+    bool initialized();
+    QString errorString();
+    QString debugInfo();
+}; // namespace Crypto
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/crypto/Crypto.h:29]: Our formatting causes the namespace to auto indent even when unindented manually.

File: src/keeshare/KeeShareSettings.cpp (hunk 3 above)
Comment (droidmonkey) [reply src/keeshare/KeeShareSettings.cpp:126]: We are going to be removing all of this code in the near future by removing signed keeshare container support. It is totally unnecessary. Recommend just keeping it as-is for this PR.

File: src/keys/FileKey.cpp (hunk 10 above)
Comment (droidmonkey) [reply src/keys/FileKey.cpp:392]: It will just be easier to do this when we implement a secure container that bridges between Botan::secure_vector and QByteArray. We use several convenience calls in this class like fromHex, fromBase64, toUpper, toLower, etc. I replaced with secure_vector where I could for now.

File: src/keys/YkChallengeResponseKey.cpp (hunk 13 above)
Comment (droidmonkey) [reply src/keys/YkChallengeResponseKey.cpp:52]: Excellent, just used m_key directly!

File: tests/TestSignature.cpp (hunk 17 above)
Comment (droidmonkey) [reply tests/TestSignature.cpp:123]: Since this is a helper function I cannot do that, but added a QVERIFY statement below each call to this function.

File: src/keys/FileKey.cpp (hunk 7 above)
Comment (droidmonkey) [reply src/keys/FileKey.cpp:215]: I added `--disable-modules=locking_allocator` to our ASAN build flags. I think that is what you meant? Botan itself adds that flag when running sanitization of the internal library components.

**Comment by droidmonkey** [comment]
@phoerious ready for re-review. Left some comments unresolved for your acceptance.

File: src/keys/FileKey.cpp (hunk 7 above)
Comment (phoerious) [reply src/keys/FileKey.cpp:212]: Yes.

File: src/keys/FileKey.cpp (hunk 10 above)
Comment (phoerious) [reply src/keys/FileKey.cpp]: Sounds good.

**Review by phoerious — APPROVED** [review]
I guess this is good to go once CI is green, but you didn't unindent the namespaces?

**Comment by droidmonkey** [comment]
I can't unindent them, at least not locally, every time I hit format they re-indent. Do we need to update our format file?

**Comment by droidmonkey** [comment]
We are now a Botan family!
